zebra-network/src/peer_set/candidate_set.rs

```diff
@@ -188,13 +203,28 @@ where
     ///
     /// Live `Responded` peers will stay live if they keep responding, or
     /// become a reconnection candidate if they stop responding.
-    pub fn next(&mut self) -> Option<MetaAddr> {
-        let mut peer_set_guard = self.peer_set.lock().unwrap();
-        let mut reconnect = peer_set_guard.reconnection_peers().next()?;
+    ///
+    /// ## Security
+    ///
+    /// Zebra resists distributed denial of service attacks by making sure that
+    /// new peer connections are initiated at least
+    /// `MIN_PEER_CONNECTION_INTERVAL` apart.
+    pub async fn next(&mut self) -> Option<MetaAddr> {
+        let now = Instant::now();
+        let mut sleep = sleep_until(now + Self::MIN_PEER_CONNECTION_INTERVAL);
+        mem::swap(&mut self.sleep, &mut sleep);
+
+        let reconnect = {
+            let mut peer_set_guard = self.peer_set.lock().unwrap();
+            let mut reconnect = peer_set_guard.reconnection_peers().next()?;
```

Comment:
This early exit is ok, because we're not actually returning a peer.

Maybe we should document that?

Reply:
I do not feel like we should be documenting early returns unless we have a specific reason to think not documenting it might exacerbate some sort of foot-gun.

Reply:
This entire function is now security-critical code, because it must ensure that new peers are yielded at least 0.1 seconds apart.

The early return skips the sleep.await that ensures this security property, but that's ok, because it's returning None, so there will be no connection in response to the early return.

So I think that we should document that:
- [ ] the sleep.await is security-critical
- [ ] the early return as a valid exception to the rule that the function must delay before returning

I'd also like us to move the sleep_until below the sleep.await, so that:
* both timers aren't running while we wait for the peer set lock to become available, while we're finding the next peer, and while sleep.await is running
* we only reset the timer when we've returned Some(peer)

This is a security-critical bug, because running both timers at the same time can halve the interval between returned peers.

Reply:
To be clear, it's the now() call that needs to move below the await.

Reply:
The change for moving the `sleep_until` isn't possible with the current code structure so I did the closest change I could without significantly reworking the code being discussed.